src/test/ui/issue-45697.stderr

```diff
@@ -0,0 +1,18 @@
+error[E0506]: cannot assign to `*y.pointer` because it is borrowed (Ast)
+  --> $DIR/issue-45697.rs:30:9
+   |
+29 |         let z = copy_borrowed_ptr(&mut y);
+   |                                        - borrow of `*y.pointer` occurs here
+30 |         *y.pointer += 1;
+   |         ^^^^^^^^^^^^^^^ assignment to borrowed `*y.pointer` occurs here
+
+error[E0506]: cannot assign to `*y.pointer` because it is borrowed (Mir)
+  --> $DIR/issue-45697.rs:30:9
+   |
+29 |         let z = copy_borrowed_ptr(&mut y);
+   |                                   ------ borrow of `*y.pointer` occurs here
+30 |         *y.pointer += 1;
+   |         ^^^^^^^^^^^^^^^ assignment to borrowed `*y.pointer` occurs here
```

Comment:
The borrow checker says it's a "use", not "assignment".

```
[00:55:48] failures:
[00:55:48] 
[00:55:48] ---- [ui] ui/issue-45697.rs stdout ----
[00:55:48] 	diff of stderr:
[00:55:48] 
[00:55:48] 6	30 |         *y.pointer += 1;
[00:55:48] 7	   |         ^^^^^^^^^^^^^^^ assignment to borrowed `*y.pointer` occurs here
[00:55:48] 8	
[00:55:48] +	error[E0503]: cannot use `*y.pointer` because it was mutably borrowed (Mir)
[00:55:48] -	error[E0506]: cannot assign to `*y.pointer` because it is borrowed (Mir)
[00:55:48] 10	  --> $DIR/issue-45697.rs:30:9
[00:55:48] 11	   |
[00:55:48] 12	29 |         let z = copy_borrowed_ptr(&mut y);
[00:55:48] 
[00:55:48] +	   |                                   ------ borrow of `y` occurs here
[00:55:48] -	   |                                   ------ borrow of `*y.pointer` occurs here
[00:55:48] 14	30 |         *y.pointer += 1;
[00:55:48] +	   |         ^^^^^^^^^^^^^^^ use of borrowed `y`
[00:55:48] -	   |         ^^^^^^^^^^^^^^^ assignment to borrowed `*y.pointer` occurs here
[00:55:48] 16	
[00:55:48] 17	error: aborting due to 2 previous errors
[00:55:48] 18	
```

Reply:
Hmm, seems like the issue we saw in the rollup again. I'll rebase it and take a look.

Reply:
Alright, so after a healthy amount of confusion and recompiling, I've made a little progress - it seems my changes simultaneously work and don't work:

Reply:
Alright, so after a healthy amount of confusion and recompiling, I've made a little progress - it seems my changes simultaneously work and don't work:

```
/mnt/c/Users/David/Projects/personal/rust issue-45697* ↓↑
❯ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustc src/test/ui/issue-45697.rs -Z borrowck=compare -Z emit-end-regions                    error[E0506]: cannot assign to `*y.pointer` because it is borrowed (Ast)
  --> src/test/ui/issue-45697.rs:30:9
   |
29 |         let z = copy_borrowed_ptr(&mut y);
   |                                        - borrow of `*y.pointer` occurs here
30 |         *y.pointer += 1;
   |         ^^^^^^^^^^^^^^^ assignment to borrowed `*y.pointer` occurs here

error[E0503]: cannot use `*y.pointer` because it was mutably borrowed (Mir)
  --> src/test/ui/issue-45697.rs:30:9
   |
29 |         let z = copy_borrowed_ptr(&mut y);
   |                                   ------ borrow of `y` occurs here
30 |         *y.pointer += 1;
   |         ^^^^^^^^^^^^^^^ use of borrowed `y`

error: aborting due to 2 previous errors


/mnt/c/Users/David/Projects/personal/rust issue-45697* ↓↑
❯ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustc src/test/ui/issue-45697.rs -Z borrowck=compare -Z emit-end-regions -O
error[E0506]: cannot assign to `*y.pointer` because it is borrowed (Ast)
  --> src/test/ui/issue-45697.rs:30:9
   |
29 |         let z = copy_borrowed_ptr(&mut y);
   |                                        - borrow of `*y.pointer` occurs here
30 |         *y.pointer += 1;
   |         ^^^^^^^^^^^^^^^ assignment to borrowed `*y.pointer` occurs here

error[E0506]: cannot assign to `*y.pointer` because it is borrowed (Mir)
  --> src/test/ui/issue-45697.rs:30:9
   |
29 |         let z = copy_borrowed_ptr(&mut y);
   |                                   ------ borrow of `*y.pointer` occurs here
30 |         *y.pointer += 1;
   |         ^^^^^^^^^^^^^^^ assignment to borrowed `*y.pointer` occurs here

error: aborting due to 2 previous errors
```

It seems that when the `-O` flag is passed (as it is with `compiletest`), the write pass happens first in the MIR borrowck, but when it isn't passed, the read pass happens first. It's the write pass error we want, so that's a problem. Now, I've got to assume that when the tests are run for a merge, `-O` isn't passed since that's the only thing I've found that really makes a difference here.

When I was working on this PR initially, I ran all my tests through `x.py`, and therefore thought my tests passed (and they did, technically) and Travis on the PR confirms that. Not sure what the solution is here, but I'm going to keep digging.

Reply:
Tried to dig around and find out what's causing this, but struggling to make any decent progress, would appreciate any pointers as to where I should be looking - @nikomatsakis.

In the paragraphs below, I'm defining passing run as when `-O` is passed as an argument and the intended error is output. Failing runs are everything else.

So, on the passing runs, the first error is from a `Shallow/Write(Mutate)` call to `access_place` from `mutate_place` which was called by `visit_statement_entry`. On the failing runs, we see a `Deep/Read(Copy)` call from the same places.

For the second error, on passing runs, we see a `Deep/Read(Copy)` run from `access_place <- consume_operator <- visit_terminator_entry`. On failing runs, we see a `Shallow/Write(Mutate)` call from `access_place <- mutate_place <- visit_statement_entry`.

Logs from the [passing run](https://gist.github.com/davidtwco/3b96611cec868212607545ac728589ea) and logs from the [failing run](https://gist.github.com/davidtwco/995f23775ca1cea62ae6c8e00e1f2c9b). These logs include some logging statements I've added to help me trace the call stack back.

As best I can tell, the issue isn't in the `borrow_check` module but in the statements/terminator that is provided to it from the `dataflow` or `transform` module differing when `-O` is passed and otherwise.